lib/tools/system-calls.js

```diff
@@ -216,4 +217,178 @@ systemCallMethods.getPortFromEmulatorString = function (emStr) {
   return false;
 };
 
+systemCallMethods.getConnectedEmulators = async function () {
+  try {
+    log.debug("Getting connected emulators");
+    let devices = await this.getConnectedDevices();
+    let emulators = [];
+    for (let device of devices) {
+      let port = this.getPortFromEmulatorString(device.udid);
+      if (port) {
+        device.port = port;
+        emulators.push(device);
+      }
+    }
+    log.debug(`${emulators.length} emulator(s) connected`);
+    return emulators;
+  } catch (e) {
+    log.errorAndThrow(`Error getting emulators. Original error: ${e.message}`);
+  }
+};
+
+systemCallMethods.setEmulatorPort = function (emPort) {
+  this.emulatorPort = emPort;
+};
+
+systemCallMethods.setDeviceId = function (deviceId) {
+  log.debug(`Setting device id to ${deviceId}`);
+  this.curDeviceId = deviceId;
+  this.adb.defaultArgs.push("-s", deviceId);
+};
+
+systemCallMethods.getRunningAVD = async function (avdName) {
+  try {
+    log.debug(`Trying to find ${avdName} emulator`);
+    let emulators = await this.getConnectedEmulators();
+    if (emulators.length < 1) {
+      log.errorAndThrow("No emulators connected");
+    }
+    for (let emulator of emulators) {
+      this.setEmulatorPort(emulator.port);
+      let runningAVDName = await this.sendTelnetCommand("avd name");
+      if (avdName === runningAVDName) {
+        log.debug(`Found emulator ${avdName} in port ${emulator.port}`);
+        this.setDeviceId(emulator.udid);
+        return emulator;
+      }
+    }
+    log.debug(`Emulator ${avdName} not running`);
+    return null;
+  } catch (e) {
+    log.errorAndThrow(`Error getting AVD. Original error: ${e.message}`);
+  }
+};
+
+systemCallMethods.getRunningAVDWithRetry = async function (avdName, timeoutMs = 20000) {
+  try {
+    let start = Date.now();
+    while ((Date.now() - start) < timeoutMs) {
+      try {
+        let runningAVD = await this.getRunningAVD(avdName.replace('@', ''));
+        if (runningAVD) {
+          return runningAVD;
+        }
+      } catch (e) {
+        // Do nothing.
+        log.info('Trying to get running AVD');
+      }
+      // cool down
+      await sleep(200);
+    }
+    log.errorAndThrow(`Could not find ${avdName} emulator.`);
+  } catch (e) {
+    log.errorAndThrow(`Error getting AVD with retry. Original error: ${e.message}`);
+  }
+};
+
+systemCallMethods.killAllEmulators = async function () {
+  let cmd, args;
+  if (system.isWindows()) {
+    cmd = 'TASKKILL';
+    args = ['TASKKILL' ,'/IM', 'emulator.exe'];
+  } else {
+    cmd = '/usr/bin/killall';
+    args = ['-m', 'emulator*'];
+  }
+  try {
+    await exec(cmd, args);
+  } catch (e) {
+    log.errorAndThrow(`Error killing emulators. Original error: ${e.message}`);
```

Comment:
the other thing I don't like about this pattern is we lose the original error stack. If `exec` was a deep complicated function, all we would get here is the error message, not a stack of exactly where in `exec` it went wrong.

so i'm kind of on the fence. what do you think @sebv?